### DISCUSSION TITLE
How to renew letsencrypt SSL cert?

### DISCUSSION THREAD

**Discussion by Computroniks** [modoboa/modoboa#2108] [Help] [answered]
# Impacted versions

* OS Type: Ubuntu 18.04.4 LTS
* Database Type: PostgreSQL
* Database version: 10.15
* Modoboa: 1.17.0
* installer used: Yes
* Webserver: Nginx

# Steps to reproduce
Standard install using installer but changing SSL cert from self signed to letsencrypt.

I have tried a manually running the command in /etc/cron.d/letsencrypt but recived the following error

```
Attempting to renew cert (mail.example.com) from /etc/letsencrypt/renewal/mail.examlpe.com.conf produced an unexpected error: Failed authorization procedure. mail.example.com (http-01): urn:ietf:params:acme:error:unauthorized :: The client lacks sufficient authorization :: Invalid response from https://mail.example.com/.well-known/acme-challenge/j2JLqbhZ0r-rLL6cjf36gDFnkQz0qcNRsIQ6Fnb0nzs [79.66.236.109]: "<!DOCTYPE html>\n\n<html lang=\"en\">\n  <head>\n    <meta charset=\"utf-8\" />\n    <title></title>\n    <link rel=\"icon\" type=\"image/png". Skipping.
All renewal attempts failed. The following certs could not be renewed:
  /etc/letsencrypt/live/mail.example.com/fullchain.pem (failure)
```

Most of the slightly older issues to do with updating the SSL cert such as #1546 and [#116 on modoboa-installer](https://github.com/modoboa/modoboa-installer/issues/116) suggest using:
```
/opt/certbot-auto -q renew --pre-hook 'nginx -t && service stop nginx stop' --post-hook 'nginx -t && service nginx start && service postfix reload && service dovecot reload'
```
or 
```
/opt/certbot-auto renew --no-self-upgrade --force-renewal
```
but it seems that /opt/certbot-auto doesn't exist. Presumably because it has been depreciated by letsencrypt. There also doesn't seem to be anything in the docs about renewing the certificate.

**Comment by MrGeneration** [comment]
According to this class it should be installed automatically:
https://github.com/modoboa/modoboa-installer/blob/master/modoboa_installer/ssl.py#L63-L110

In my humble opinion generation of the initial certificate should have failed if that was not the case. Possibly it was removed too.
It's hard to tell without knowing which command you're running.

A quick and dirty try may be to manually install certbot ( https://certbot.eff.org/instructions ) and try again.

**Reply by Computroniks** [reply]
> According to this class it should be installed automatically:
> https://github.com/modoboa/modoboa-installer/blob/master/modoboa_installer/ssl.py#L63-L110
> 
> In my humble opinion generation of the initial certificate should have failed if that was not the case. Possibly it was removed too.
> It's hard to tell without knowing which command you're running.
> 
> A quick and dirty try may be to manually install certbot ( https://certbot.eff.org/instructions ) and try again.

I tried installing certbot manually and ran `sudo certbot renew --dry-run` but only to receive the same error. I assume that that class renews the cert when it has 30 days until expiry.

**Reply by MrGeneration** [reply]
You may want to try the `--debug` switch to learn more.
(see: https://certbot.eff.org/docs/using.html )

**Reply by Computroniks** [reply]
It seems that when running somthing like `certbot renew`, certbot can't get access to the webserver to place the verification file. When certbot trys to view the file all it gets is 
```
<!DOCTYPE html>\n\n<html lang=\"en\">\n  <head>\n    <meta charset=\"utf-8\" />\n    <title></title>\n    <link rel=\"icon\" type=\"image/png"
```
There are also some tracebacks in the log file related to the same issue of incorrect verification file but the rest of it seems to be just printing out nginx.conf and other conf files.

**Reply by Computroniks** [reply]
Actually. I just saw somthing in the log. It seems that when certbot is trying to verify the file it gets a 403 error. I'm not sure why though.

**Reply by Computroniks** [reply]
Also I tried stopping nginx and running certbot with the standalone flag. This just results in a 400 error.

**Comment by gawel** [comment]
You're trying to renew a certificat for a domain that you don't own (mail.example.com)
I guess you leaved defaults values somewhere in the config file

**Reply by Computroniks** [reply]
> You're trying to renew a certificat for a domain that you don't own (mail.example.com)
> I guess you leaved defaults values somewhere in the config file

No that is just me not wanting to share my domain name

**Comment by Computroniks** [comment, marked as answer]
One way to update the ssl cert is by following this guide. https://github.com/modoboa/modoboa-installer/issues/16#issuecomment-244417643 Note that you do need to add the line to the server on 443 as well as 80.

**Comment by CyberfoxDK** [comment]
Had similar issues renewint letsencrypt, and it seems like the certbot nginx plugin was missing in my modoboa version. Followed this topic, and it worked right out of the box after then.

https://github.com/modoboa/modoboa/issues/2129
